### ISSUE TITLE
Shader UV calculated wrong when using regions, calculated really wrong when using .tres files and there seems to be an alpha problem with 3.4.x

### ISSUE THREAD

**Issue by chucklepie** [godotengine/godot#57401]
### Godot version

3.4.2

### System information

PopOS

### Issue description

I don't know if this needs to be 2 or 3 separate bugs, but I'll put them all down here:

Simple shader that transforms a sprite using a gradient texture. Three images:

1. A single image 80x64, this is the image on the right and works
2. A sprite 240x128 with 8 images but a region specifying a single sprite, this is the image in the middle
3. A sprite defined exactly as (2) but as a .tres file from a larger sprite.

Open the node file if not open, select any image (I have shared resources), go to the material and slide the dissolve value

Bug 1: the texture/UV calculations seem to be taking into the account the entire size of the source image, so if you look image 3 (the single png) works correctly matching the texture, but image 2 and 3 do not

Bug 2: I don't know if this is just the size or that one is a tres file, but while images 2 and 3 finish at the same time, the .tres file finishes rapidly and cannot be revealed smoothly.

Bug 3: Maybe it's my code but if you look, the images should be completely transparent until the dissolve is finished. In Godot 3.3 this was the case but now they are not working properly. If you edit the shader file and change the alpha decrement you'll see it seems to finding colours where there are none.

https://user-images.githubusercontent.com/12863685/151662215-5f161ebe-9d4a-433f-85f1-e798a054e9f1.mp4

If these are known bugs or valid bugs, I'd appreciate a workaround if any.

### Steps to reproduce

As description.

### Minimal reproduction project

[tres_shader_fail.zip](https://github.com/godotengine/godot/files/7963748/tres_shader_fail.zip)

**Comment by chucklepie** [comment]
Just a note, that weird colouring you can see (my bug 3) does not happen with the default icon.png. But this is just an ordinary RGBA8 png?

**Comment by chucklepie** [comment]
Hello again,
I think something is very, very wrong with 3.4.2 and it looks like it is related to the import process, regarding my bug 3, when using any shader:

This is the normal image, it is using the 2d Pixel template (i.e. filter is off).

![Screenshot 2022-01-29 20:19:45](https://user-images.githubusercontent.com/12863685/151676329-f932f5c1-3593-4f62-bc7d-a50c22bbae02.png)

When I turn on my shader, any shader, which literally does nothing:

COLOR=texture(TEXTURE,UV);

This is what it looks like:

![Screenshot 2022-01-29 20:21:47](https://user-images.githubusercontent.com/12863685/151676384-347a2b84-a47e-4282-bbbc-b8fad2a2597b.png)

If I then disable fix_alpha_border, it looks like this:

![Screenshot 2022-01-29 20:23:35](https://user-images.githubusercontent.com/12863685/151676418-8ad3beec-1cf1-4618-bbc7-5eb87a0c2066.png)

And with fix_alpha_border off as above, it does this to the textures that atlases:
![Screenshot 2022-01-29 20:24:45](https://user-images.githubusercontent.com/12863685/151676463-93369c99-40df-4254-9044-8fbc84b33884.png)

**Comment by Calinou** [comment]
Can you reproduce this after enabling **Force Png** in the project settings, exiting the editor, removing the `.import/` folder within the project then opening the editor again?

**Comment by chucklepie** [comment]
I'll try now... hang on, but have you tried running the project?

**Comment by belzecue** [comment]
With 3.5.1-stable, today I was trying to apply a UV box mask via a canvas_item custom shader to a Sprite with region and I came across this same bug.

Another repro:

1. Set up a Sprite region from an atlas texture, e.g. isolate an object from any of Kenny's sprite atlases.
2. Swap in a custom shader.
3. Use this shader code:

shader_type canvas_item;

```
void fragment() {
	COLOR = texture(TEXTURE, UV * vec2(1.0));
}

```
4. Observe that the correctly regioned sprite is displaying.

![image](https://user-images.githubusercontent.com/1931303/209665341-bf00abdc-a44f-4d23-9dd0-bc928fec1005.png)

6. Change the 1.0 to 1.2, i.e. zoom out 20%.
7. Observe that there are other atlas tiles now showing in the region.  Godot is using the entire atlas UV space.

![image](https://user-images.githubusercontent.com/1931303/209665415-c49949f4-4944-4e72-ac00-ddfba41c0009.png)

So I guess this means you cannot do UV manipulation in custom shaders when using a Sprite with a defined region.

**Comment by dsmtE** [comment]
I also needed to apply a relative effect to my sprite (pixelate effect).
But it's not a bug, the UV currently received refer to the underlying texture.

So we have to get somehow the information of the currently set region to be able to make the necessary transformations to get "local" region_UV.

Here are two functions that allow to do this transformation:

```glsl
vec2 uv_to_region_uv(vec2 uv, vec2 texture_pixel_size) {
	return (uv-(rect_info.xy*texture_pixel_size))/(rect_info.zw*texture_pixel_size);
}

vec2 region_uv_to_uv(vec2 region_uv, vec2 texture_pixel_size) {
	return (rect_info.xy*texture_pixel_size)+region_uv*(rect_info.zw*texture_pixel_size);
}

void fragment() {
	vec2 region_uv = uv_to_region_uv(UV, TEXTURE_PIXEL_SIZE);
	// Any local uv transformation
	vec2 pixelated_region_uv = pixalate(region_uv, rect_info.zw)+TEXTURE_PIXEL_SIZE/2.;
	// Convert back to UV texture to sample it
	vec2 texture_uv = region_uv_to_uv(pixelated_region_uv, TEXTURE_PIXEL_SIZE);
	COLOR = texture(TEXTURE, texture_uv);
}
```
Here, rect_info contains the position and size info of the currently set region.

Here I set manually the uniform rect_info, but I had the same problem in Unity and I made a dirty script that allows to update the uniform automatically.

```c#
using System;

using UnityEngine;

[ExecuteInEditMode, RequireComponent(typeof(SpriteRenderer))]
public class SpriteRectUniformProvider : MonoBehaviour 
{
    // (...)
    public void UpdateMaterialPropertyBlock() {
        Sprite sprite = _spriteRenderer.sprite;
        if (sprite != null)
        {
            MaterialPropertyBlock mpb = new MaterialPropertyBlock();
            _spriteRenderer.GetPropertyBlock(mpb);
            
            Vector4 spriteRect = new Vector4(
                sprite.rect.xMin / sprite.texture.width,
                sprite.rect.yMin / sprite.texture.height,
                sprite.rect.width,
                sprite.rect.height);
            
            mpb.SetVector("_spriteRect", spriteRect);
            _spriteRenderer.SetPropertyBlock(mpb);
        }
        else
        {
            string logMessage = $"[{GetType().Name}] Unable to update uniform as the sprite isn't set yet on the SpriteRenderer";
            if (Application.isPlaying)
            {
                JdLog.LogThat("", LogLevel.LevelInfo, logMessage);
            }
            else
            {
                Debug.Log(logMessage);
            }
        }
    }
}
```

I don't know if it's useful to have the UV of the underlying texture (when using a region) but if it's the case (so you can't just change the UV), a great idea would be to **have a uniform build-in variable to get the region info if needed**.

I started looking into the engine (open source yeah), I would like to make a PR to bring this functionality. Do you think this is the right direction (before I start anything), or do you have any other ideas for a solution in this situation ?

**Comment by dsmtE** [comment]
@Calinou do you have an opinion ? 
I search in those files :
- [godot](https://github.com/godotengine/godot/tree/18a2e7ff6ebf0a477138c12c3efa63c15270a2ee)/[servers](https://github.com/godotengine/godot/tree/18a2e7ff6ebf0a477138c12c3efa63c15270a2ee/servers)/[rendering](https://github.com/godotengine/godot/tree/18a2e7ff6ebf0a477138c12c3efa63c15270a2ee/servers/rendering)/shader_language.cpp
- [godot](https://github.com/godotengine/godot/tree/18a2e7ff6ebf0a477138c12c3efa63c15270a2ee)/[servers](https://github.com/godotengine/godot/tree/18a2e7ff6ebf0a477138c12c3efa63c15270a2ee/servers)/[rendering](https://github.com/godotengine/godot/tree/18a2e7ff6ebf0a477138c12c3efa63c15270a2ee/servers/rendering)/shader_compiler.cpp
Do you have some tips to allow me to implement built-in new uniform (region_rect) for the canvas_item shader_type ?

**Comment by Mickeon** [comment]
I believe I'm currently experiencing this same issue in 4.1 and it's quite the burden. Is there any decently viable workaround? Can this even be fixed someday?

**Comment by dsmtE** [comment]
Hello,
@Mickeon, you can use my workaround using unction **uv_to_region_uv** and **region_uv_to_uv** above.

I proposed to fix it myself (or make a uniform build-in) to manage this case, but I didn't get any feedback from the godot team before starting a PR and working on it.
@Calinou ?

**Comment by Calinou** [comment]
@dsmtE Feel free to open a pull request for this :slightly_smiling_face:

**Comment by GrauBlitz** [comment]
I would thank you for a PR as this is still present in 4.2.1-stable 🙂

**Comment by dsmtE** [comment]
Hello @Calinou,
I'm starting to work on it, it's a bit difficult to understand and integrate into the code base. First, I'll try to make it work in the dirtiest way and then we'll rework it and do it the right way.

I found the glsl canvas_item shader (./servers/rendering/renderer_rd/shaders/canvas.glsl) and tried to figure out where the sprite region information (offset and size) came from.

Here's a list of several parts of the code that I found interesting:
- ./servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp(method "RendererCanvasRenderRD::_render_item", line 508)
- ./scene/2d/sprite_2d.cpp(method "Sprite2D::set_region_enabled", line 212)
- .//scene/resources/texture.cpp(method "Texture2D::draw_rect_region", line 77)

I managed to add a custom variable to make certain information accessible from my Godot shader (draw_data.src_rect and draw_data.dst_rect)
```glsl
vec4 region_position_and_size = draw_data.dst_rect;
```

According to my tests, **raw_data.dst_rect.zw** seems to match the region size.
And so I'm able to perform the desired scale transformation on my UVs when the region has an offset of 0.

 But, I can't figure out how to get the position offset. I can't figure out what **raw_data.dst_rect.xy** is. 
Can you enlighten me a little or ask someone who has information on this subject so that I can continue? 

Thanks in advance 🙂

**Comment by Delsin-Yu** [comment]
> Hello @Calinou, I'm starting to work on it, it's a bit difficult to understand and integrate into the code base. First, I'll try to make it work in the dirtiest way and then we'll rework it and do it the right way.
> 
> I found the glsl canvas_item shader (./servers/rendering/renderer_rd/shaders/canvas.glsl) and tried to figure out where the sprite region information (offset and size) came from.
> 
> Here's a list of several parts of the code that I found interesting:
> 
> * ./servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp(method "RendererCanvasRenderRD::_render_item", line 508)
> * ./scene/2d/sprite_2d.cpp(method "Sprite2D::set_region_enabled", line 212)
> * .//scene/resources/texture.cpp(method "Texture2D::draw_rect_region", line 77)
> 
> I managed to add a custom variable to make certain information accessible from my Godot shader (draw_data.src_rect and draw_data.dst_rect)
> 
> ```glsl
> vec4 region_position_and_size = draw_data.dst_rect;
> ```
> 
> According to my tests, **raw_data.dst_rect.zw** seems to match the region size. And so I'm able to perform the desired scale transformation on my UVs when the region has an offset of 0.
> 
> But, I can't figure out how to get the position offset. I can't figure out what **raw_data.dst_rect.xy** is. Can you enlighten me a little or ask someone who has information on this subject so that I can continue?
> 
> Thanks in advance 🙂

Is it possible to share your custom build of the Engine? Would love to help.

**Comment by dsmtE** [comment]
Hello,
Thanks for your help @Delsin-Yu!
I have paused my work on this for a few weeks (for personal reasons).
This week, with the help of [this](https://chat.godotengine.org/channel/new-contributors?msg=Sg9ggP3emTQNRExze) message by @kleonc (on the Godot contributor chat), I've figured out how to use `draw_data.src_rect` for my needs.

I still don't understand all the shader variants and defines (USE_PRIMITIVE, USE_ATTRIBUTES, USE_NINEPATCH), I would be happy to discuss it with you if you can explain it to me (anyway there will be reviews and exchanges after my first "fix" proposal because I didn't manage the **gles3** part for the moment to make my life easier).

I will iterate a bit more this week and come back with some proposals and examples here on the possibilities of how this fix should be properly integrated into the engine (in a backward-compatible way or not, change the management of the existing UV variable or just make available new built-in functions and variables). Furthermore, I will also give you a link to the branch I'm working on at the same time.

**Comment by dsmtE** [comment]
Hello,
I have a first working draft :confetti_ball: !

Basically, I added two new built-in variables and two new  built-in functions:
- `REGION_SIZE` : (vec2) which contains the size of the region set in the inspector (in pixels)
- `REGION_POSITION` :  (vec2) which contains the position of the region (in pixels)
- `region_to_texture_uv(vec2 uv)`: that take uv (from range [0, 1] and transform them to scan and shift to the region location
```glsl
vec2 region_to_texture_uv(vec2 uv) {
    return uv * region_info.zw + region_info.xy;
}
```
- `texture_to_region_uv(vec2 uv)`: that do the reverse transformation to get normalized [0,1] uv coordinates from the applied region uv
```glsl
vec2 texture_to_region_uv(vec2 uv) {
    return (uv - region_info.xy) / region_info.zw;
}
```

These built-in allows us to do local transformation on the region sub-texture.
Here is an example of shader using these built-in to zoom in :

```glsl
shader_type canvas_item;

void fragment() {
    // Convert texture space uv to region
    vec2 region_uv = texture_to_region_uv(UV);

    // Any local uv transformation : exemple with centered zoom effect
    float zoom_factor = 1.2;
    vec2 zoomed_uv = (region_uv-0.5)/zoom_factor+0.5;

    // Convert back to texture uv to be sampled properly using the texture(TEXTURE, uv) function
    vec2 texture_uv = region_to_texture_uv(zoomed_uv);

    // Sample the texture using right region UV after local transformation
    COLOR = texture(TEXTURE, texture_uv);
}
```

Doing it this way do not break any older shader (as the built-in UV variable isn't changed).

Another benefit of doing it that way is to be able to manipulate also the texture UV after the "back conversion" from local to texture uv.

Here an example of shifting the UV by the size of the region (could be useful for sprite Sheet effect when the sprite2D use region already):

```glsl
shader_type canvas_item;

void fragment() {
    vec2 texture_uv = UV;
	
    // Shift the UV on the x axis using REGION_SIZE to obtain the new part of the texture 
    // with the same size as the region (next texture in a horizontal sheet sprite, for example)

    // This must be multiplied by TEXTURE_PIXEL_SIZE 
    //because REGION_SIZE contains the size of the region in pixels (16x16 for example) and to offset the uvs we need to be in normalized uv space
    texture_uv.x += REGION_SIZE.x*TEXTURE_PIXEL_SIZE.x;
    COLOR = texture(TEXTURE, texture_uv);
}
```

What do you think of this proposal (from a technical and user point of view)?
@Delsin-Yu, @belzecue, @Mickeon, @chucklepie, @GrauBlitz

---

I prefer this approach because it doesn't break backwards compatibility, but allows you to do whatever you want in custom shaders with regions.

However, I wonder if we shouldn't change what the UV variable itself contains. After all, the basic behavior of regions (i.e. texture UVs for sampling and not local UVs [0,1] for the region before sampling) may not be the most intuitive in consideration of the preceding [messages](https://github.com/godotengine/godot/issues/57401#issuecomment-1365864349).

We could then imagine a possible solution that :
- change the content of UV to contain local UV instead (equivalent to region_uv in my previous example)
- propose a function for sampling in a texture and applying the UV change for the region at that moment, or expose the region_to_texture_uv function but make it necessary before sampling to display what you want.

An example of simple fragment shader become:
```glsl
shader_type canvas_item;

void fragment() {
    COLOR = texture(TEXTURE, region_to_texture_uv(UV));
}
```
and with zoom effect:
```glsl
shader_type canvas_item;

void fragment() {
    vec2 zoomed_uv = (UV-0.5)/zoom_factor+0.5;
    COLOR = texture(TEXTURE, region_to_texture_uv(zoomed_uv));
}
```

This solution breaks backward compatibility, as it requires the transformation function to be called before sampling.

An even "smoother" solution would be to change the behavior of the `texture(TEXTURE, UV)` sampling function itself to perform the region transformation implicitly.

Example:
```glsl
shader_type canvas_item;

void fragment() {
    // implicit region UV transformation here
    COLOR = texture(TEXTURE, UV * 1.2);
}
```

But this would prevent the actual texture UVs from being modified as freely as we want, as in my previous example of shifting region effect, where the shift must be performed on the texture UVs and not on the UVs local to the region (so after the potential implicit region_to_texture_uv transformation).

And I don't know how to do it technically either at the moment, another reason not to go for this solution. :smiling_face_with_tear: 

---

Here my repo and branch I'm working on: https://github.com/dsmtE/godot/tree/dev/region_texture_mapping
Reminder, this is a rough version and doesn't support gles3 and probably other things to be integrated properly into the engine.  I prefer to discuss it and get your feedbacks before continuing my work.

**Comment by Delsin-Yu** [comment]
@dsmtE Wow, that's a good job you have done, congrats!
I think you can maybe even create a draft PR for this.
---
I prefer the `REGION_SIZE`/`REGION_POSITION`+`region_to_texture_uv`+`texture_to_region_uv` approach as well since it gives full control to the developer (not to mention it doesn't impact existing shader code, so it's not that hard to get merged).

I also like the idea of silently changing the `texture(TEXTURE, UV)` behavior so it kind of silently addresses the issue for future developers. But this breaks existing shaders that use uniform `region size`/`region position` to perform manual correction.

**Comment by Mickeon** [comment]
@dsmtE I would suggest directly opening a PR for this instead of talking about it here, to make it more focused. I have no opinions right now.

**Comment by maidopi-usagi** [comment]
Glad to see if built-in variables or functions could be exposed, as it will be much more convenient to use~

Before a new PR got merged, this following piece of shader may be a bit helpful if someone is looking for a quick fix. It might be kind of dirty and not optimized, but it just works.

```glsl
shader_type canvas_item;

const vec2 VERTICES[4] = vec2[](vec2(0.0),vec2(0.0,1.0),vec2(1.0),vec2(1.0,0.0));
varying vec2 originUV;
varying flat vec4 regionInfo;

void vertex() {
	originUV = VERTICES[VERTEX_ID];
	vec2 ouv = VERTICES[0];
	regionInfo.xy = VERTEX.xy/(ouv - 0.5);
	regionInfo.zw = UV;
}

vec2 region_from_texture(vec2 uv, vec2 size) {
	return  regionInfo.zw + uv / size * regionInfo.xy;
}

vec2 texture_from_region(vec2 uv, vec2 size) {
	return uv * size / regionInfo.xy - regionInfo.zw;
}

void fragment() {
	vec2 uv_to_transform = (originUV - 0.5) * sin(TIME) * 3.0 + 0.5;

	const vec2 size = vec2(textureSize(TEXTURE, 0));
	vec2 restoredUV = region_from_texture(uv_to_transform,size);

	COLOR = texture(TEXTURE, restoredUV);
}
```

https://github.com/godotengine/godot/assets/33864304/3c26d862-3a6e-43be-8357-d1a4c96a121a

**Comment by kleonc** [comment]
@dsmtE AFAICT the rendering team don't like to break previously working shaders, so I'd guess that auto-changing `UV` to mean different thing (depending on the context) than it currently does is a no-no. Same about changing the behavior of already existing `texture` function. So I suggest focusing on non-compat-breaking solutions.

What could probably be considered as a viable solution for changing what the built-in `UV` refers to is adding a new [render mode](https://docs.godotengine.org/en/4.2/tutorials/shaders/shader_reference/canvas_item_shader.html#render-modes). Just as currently e.g. `world_vertex_coords` can be used to change what `VERTEX` within `vertex()` is, something like `region_uv_coords` could maybe be added for changing what `UV` refers to? :thinking:

But regardless, I'd say instead of commenting here you should probably open a proposal in https://github.com/godotengine/godot-proposals, that's exactly what that repo is for - there's some issue and you're proposing a specific solution(s) for it. Then it can be discussed in detail in there (technical details, different approaches, etc.).

**Comment by dsmtE** [comment]
Thanks for all your answers, I just open the proposal to discuss it, and after, I will continue to work on my PR.
